mpire/comms.py

```diff
@@ -132,29 +132,29 @@ def init_comms(self) -> None:
         # Task related
         self._task_queues = [self.ctx.JoinableQueue() for _ in range(self.n_jobs)]
         self._worker_running_task_locks = [self.ctx.RLock() for _ in range(self.n_jobs)]
-        self._worker_running_task = [self.ctx.Value('b', False, lock=False) for _ in range(self.n_jobs)]
-        self._worker_working_on_job = [self.ctx.Value('i', 0, lock=True) for _ in range(self.n_jobs)]
+        self._worker_running_task = [self.ctx.Value(ctypes.c_bool, False, lock=False) for _ in range(self.n_jobs)]
+        self._worker_working_on_job = self.ctx.Array('i', self.n_jobs, lock=True)
 
         # Results related
         self._results_queue = self.ctx.JoinableQueue()
         self._results_added = [0 for _ in range(self.n_jobs)]
-        self._results_received = [self.ctx.Value('L', 0, lock=True) for _ in range(self.n_jobs)]
+        self._results_received = self.ctx.Array('L', self.n_jobs, lock=True)
 
         # Worker status
-        self._worker_restart_array = [self.ctx.Value('b', False, lock=True) for _ in range(self.n_jobs)]
-        self._workers_dead = [self.ctx.Event() for _ in range(self.n_jobs)]
-        [worker_dead.set() for worker_dead in self._workers_dead]
-        self._workers_time_task_started = [self.ctx.Value('d', 0.0, lock=True) for _ in range(self.n_jobs * 3)]
+        self._worker_restart_array = self.ctx.Array(ctypes.c_bool, self.n_jobs, lock=True)
+        self._workers_dead = self.ctx.Array(ctypes.c_bool, self.n_jobs, lock=True)
+        self._workers_dead[:] = [True] * self.n_jobs
+        self._workers_time_task_started = [self.ctx.Array('d', 3, lock=True) for _ in range(self.n_jobs)]
```

Comment:
Wouldn't it be more efficient to use one array? (Using the same indexing scheme as before.) Or does that give problems?

Reply:
Well, the thing is that when you have one array with one lock, workers would have to wait for each other (lock has to be free). Now, each worker can put stuff in there independently, without the need to wait. 

Then again, waiting is only needed when workers are likely to finish tasks at the same time (e.g., when you have small tasks). When multiprocessing it's more likely that you're running longer tasks, so the waiting for the lock might be less of a problem. 

It's a trade-off between potentially waiting a tiny bit and having only one lock for n amount of workers, or having no waiting time but n locks. What do you think? Either way I can check whether the unit tests run less quickly when making that change

Reply:
I can write to a shared array from 10 processes a hundred times each in 65 ms, so I think it wouldn't add much overhead. The lock only has to be active a very short time, after all.
Well, unless maybe somehow a garbage collection starts right when you locked, I think that's only a risk if you allocate new object though. (And it's very infrequent anyway.)

Reply:
True. Ok, thnx. Will change it